### PR TITLE
feat: decompose

### PR DESCRIPTION
**Pull request by kobayu858**
## Description

- Computation of node deadlines for decomposition-based scheduler

## Related links

- original paper [OPEN_LINK](https://drive.google.com/file/d/1YH-MAWsw1VAvxqco4Beg3njQ9ykG-sb-/view?usp=drive_link) p6
- paper summary [EmbIV_INTERNAL_LINK](https://docs.google.com/presentation/d/1rabhhA9ziIcBjvT8XPLue78XEWcKMvH4/edit?usp=drive_link&ouid=114908281052440997053&rtpof=true&sd=true) p37
- design specifications [EmbIV INTERNAL LINK](https://docs.google.com/presentation/d/14YO4wHY1QqGOjoY9_g4djvtTaBXV0RCj/edit?usp=drive_link&ouid=114908281052440997053&rtpof=true&sd=true) p19

## Pre-review checklist for the PR author

- [x] I've confirmed the [self-checklist](https://docs.google.com/spreadsheets/d/1p5KMcm752iCySHJz0xL1qf6fcLDKprbo/edit#gid=863618628).
- [x] The PR is ready for review.

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

### CODE REVIEW

File: 2014_TPDS_basic_decomposition_based_algorithm/src/decomposition.rs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        let node_i = NodeIndex::new(i);
+        if node_deadline.fract() == 0.0 {
+            dag.add_param(node_i, "deadline_factor", 1);
+            dag.add_param(node_i, "deadline", *node_deadline as i32);
+            continue;
+        }
+
+        let rounded_node_deadline: f32 = format!("{:.5}", node_deadline).parse().unwrap();
+        let integer_part_str = rounded_node_deadline.trunc().abs().to_string();
+        let decimal_part_str = rounded_node_deadline
```
Comment (atsushi421) [inline 2014_TPDS_basic_decomposition_based_algorithm/src/decomposition.rs:30]:
```suggestion
        let fractional_part_str = rounded_node_deadline
```

File: 2014_TPDS_basic_decomposition_based_algorithm/src/decomposition.rs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        });
+    }
+
+    for (i, node_deadline) in nodes_deadline.iter().enumerate() {
+        let node_i = NodeIndex::new(i);
+        if node_deadline.fract() == 0.0 {
+            dag.add_param(node_i, "deadline_factor", 1);
+            dag.add_param(node_i, "deadline", *node_deadline as i32);
+            continue;
+        }
```
Comment (atsushi421) [inline 2014_TPDS_basic_decomposition_based_algorithm/src/decomposition.rs:26]: This process is not performed very often and can be deleted.

File: 2014_TPDS_basic_decomposition_based_algorithm/src/decomposition.rs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+            .to_string()
+            .chars()
+            .skip(integer_part_str.len() + 1)
+            .collect::<String>();
+
+        let deadline_factor = 10u64.pow(decimal_part_str.len().try_into().unwrap()) as i32;
+        dag.add_param(node_i, "deadline_factor", deadline_factor);
+        dag.add_param(
+            node_i,
+            "deadline",
```
Comment (atsushi421) [inline 2014_TPDS_basic_decomposition_based_algorithm/src/decomposition.rs:41]:
```suggestion
            "integer_scaled_deadline",
```

**Review by atsushi421 — CHANGES_REQUESTED** [review]

File: 2014_TPDS_basic_decomposition_based_algorithm/src/decomposition.rs (hunk 1 above)
Comment (kobayu858) [reply 2014_TPDS_basic_decomposition_based_algorithm/src/decomposition.rs]: Fixed.

File: 2014_TPDS_basic_decomposition_based_algorithm/src/decomposition.rs (hunk 3 above)
Comment (kobayu858) [reply 2014_TPDS_basic_decomposition_based_algorithm/src/decomposition.rs]: Changed.

**Review by atsushi421 — APPROVED** [review]
LGTM

**Pull request merged**